storage/uri.go

```diff
@@ -1,204 +1,465 @@
+// resource, except for reading it in and writing all the data back out.
+//
+// This method can fail in several ways:
+//
+// * Different permissions or credentials are required to write to the
+//   referenced resource.
+//
+// * This URI scheme could represent some resources that can be
+//   written, but this particular URI references a resources that is
+//   not something that can be written.
+//
+// * Attempting to set up the writer depended on a lower level
+//   operation such as a network or filesystem access that has failed
+//   in some way.
+//
+// * If the scheme of the given URI does not have a registered
+//   WriteableRepository instance, then this method will fail with a
+//   repository.ErrOperationNotSupported.
+//
+// Writer is backed by the repository system - this function calls into a
+// scheme-specific implementation from a registered repository.
+//
+// Since: 2.0.0
+func Writer(u fyne.URI) (fyne.URIWriteCloser, error) {
+	repo, err := repository.ForURI(u)
+	if err != nil {
+		return nil, err
+	}
 
-	} else {
-		var err error
-		parent, err = parentGeneric(s)
-		if err != nil {
-			return nil, err
-		}
+	wrepo, ok := repo.(repository.WriteableRepository)
+	if !ok {
+		return nil, repository.ErrOperationNotSupported
 	}
 
-	return NewURI(u.Scheme() + "://" + parent), nil
+	return wrepo.Writer(u)
 }
 
-// Child appends a new path element to a URI, separated by a '/' character.
+// CanWrite determines if a given URI could be written to using the Writer()
+// method. It is preferred to check if a URI is writable using this method
+// before calling Writer(), because the underlying operations required to
+// attempt to write and then report an error may be slower than the operations
+// needed to test if a URI is writable.  Keep in mind however that even if
+// CanWrite returns true, you must still do appropriate error handling for
+// Writer(), as the underlying filesystem may have changed since you called
+// CanWrite.
+
 //
-// Since: 1.4
-func Child(u fyne.URI, component string) (fyne.URI, error) {
-	// While as implemented this does not need to return an error, it is
-	// reasonable to expect that future implementations of this, especially
-	// once it gets moved into the URI interface will need to do so. This
-	// also brings it in line with Parent().
+// CanWrite is backed by the repository system - this function calls into a
+// scheme-specific implementation from a registered repository.
+//
+// Since: 2.0.0
+func CanWrite(u fyne.URI) (bool, error) {
+	repo, err := repository.ForURI(u)
+	if err != nil {
+		return false, err
+	}
 
-	s := u.String()
+	wrepo, ok := repo.(repository.WriteableRepository)
+	if !ok {
+		return false, repository.ErrOperationNotSupported
+	}
 
-	// guarantee that there will be a path separator
-	if s[len(s)-1:] != "/" {
-		s += "/"
+	return wrepo.CanWrite(u)
+}
+
+// Copy given two URIs, 'src', and 'dest' both of the same scheme , will copy
+// one to the other.
```

Comment:
Can this articulate the behaviour when the URIs don't have the same scheme?

Reply:
97c801b7da72e6bc37dd010538143712fee9147d